library/cpp/src/core/utils/profile/resource_monitor.cc

```diff
@@ -0,0 +1,75 @@
+#include "wavemap/core/utils/profile/resource_monitor.h"
+
+#include <string>
+
+namespace wavemap {
+void ResourceMonitor::start() {
+  if (running_) {
+    LOG(WARNING) << "Tried to start timer that was already running.";
+    return;
+  }
+
+  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &episode_start_cpu_time_);
+  clock_gettime(CLOCK_MONOTONIC, &episode_start_wall_time_);
+  running_ = true;
+}
+
+void ResourceMonitor::stop() {
+  if (!running_) {
+    LOG(WARNING) << "Tried to stop timer that was not running.";
+    return;
+  }
+
+  timespec episode_stop_cpu_time{};
+  timespec episode_stop_wall_time{};
+  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &episode_stop_cpu_time);
+  clock_gettime(CLOCK_MONOTONIC, &episode_stop_wall_time);
+  running_ = false;
+
+  last_episode_cpu_duration_ =
+      computeDuration(episode_start_cpu_time_, episode_stop_cpu_time);
+  last_episode_wall_duration_ =
+      computeDuration(episode_start_wall_time_, episode_stop_wall_time);
+  total_cpu_duration_ += last_episode_cpu_duration_;
+  total_wall_duration_ += last_episode_wall_duration_;
+}
+
+std::optional<size_t> ResourceMonitor::getCurrentRamUsageInKB() {
+  std::ifstream file("/proc/self/status");
```

Comment:
How do we want to handle non-linux systems with this?

Reply:
We currently don't handle them. But it will fail gracefully and return a `std::nullopt`.

Reply:
After a verbal discussion, we decided to keep it like this for now and extend support of the ResourceMonitor to other OSs when the need comes up. I will, however, add macros so that the code doesn't generate compilation errors on other platforms (probably as macros that nop out if the Linux system headers are not found).